### PR TITLE
Refactoring for grpc load balacing

### PR DESCRIPTION
**Pull request by tei-k**
# Description
https://scalar-labs.atlassian.net/browse/DLT-5555

# Done
- Remove nlb of scalardl
- Fix to using dns for grpc load balacing
- Fix to create envoy resource in default
- Fix to using v1.12.3 for envoy (fix deprecated)

### CODE REVIEW

**Comment by feeblefakie** [comment]
@tei-k @ymorimo @penpyt 
Just an idea. How about having something like `blue_enable_dns` and `green_enable_dns` and make the B/G deployment work as follows ?

1. (initial) `blue_resource_count` = 3, `blue_enable_dns` = true, `green_resource_count` = 0, `green_enable_dns` = false
1. (green up) `blue_resource_count` = 3, `blue_enable_dns` = true, `green_resource_count` = 3, `green_enable_dns` = true
1. (blue dns disabled) `blue_resource_count` = 3, `blue_enable_dns` = false, `green_resource_count` = 3, `green_enable_dns` = true
1. (blue count 0) `blue_resource_count` = 0, `blue_enable_dns` = false, `green_resource_count` = 3, `green_enable_dns` = true

Basically,  if both `blue_resource_count` is bigger than 0 and `blue_enable_dns` is true, then DNS entries for those blue nodes are created.

**Comment by tei-k** [comment]
@feeblefakie 

Thanks! 👍 
Added `blue_enable_dns ` and `green_enable_dns` flag in d99cb15.

File: modules/universal/envoy/provision/envoy.yaml [hunk 1]
```diff
@@ -59,12 +60,20 @@ static_resources:
...
+      drain_connections_on_host_removal: true
       load_assignment:
         cluster_name:              scalar-service
         endpoints:
         - lb_endpoints:
           - endpoint:
              address:
               socket_address:
-                address:           scalardl-lb.internal.scalar-labs.com
+                address:           scalardl.internal.scalar-labs.com
```
Comment (feeblefakie) [inline modules/universal/envoy/provision/envoy.yaml]: I think this shouldn't use of internal domain, right ?

File: modules/aws/scalardl/scalardl.tf [hunk 2]
```diff
@@ -149,135 +149,17 @@ resource "aws_security_group_rule" "scalardl_egress" {
...
-    evaluate_target_health = true
-  }
+  ttl     = "300"
+  records = concat(
+    local.scalardl.blue_enable_dns ? module.scalardl_blue.ip : [],
+    local.scalardl.green_enable_dns ? module.scalardl_green.ip : []
+  )
 }
 
 resource "aws_route53_record" "scalardl-blue-dns" {
```
Comment (feeblefakie) [inline modules/aws/scalardl/scalardl.tf]: I have a little concern, which I shared in a little chat with @tei-k .
`blue_enable_dns = false` sounds like we disable dns for blue cluster, but actually there are some dns entries here, so it's a little confusing.
So, either of the following might be needed.
- Renaming of the variable (to `blue_enable_dns_for_envoy` or something)
- Disable all the related DNS as well

Thought ? @tei-k @ymorimo @penpyt

File: modules/universal/envoy/provision/envoy.yaml (hunk 1 above)
Comment (tei-k) [reply modules/universal/envoy/provision/envoy.yaml]: 👍 Fixed to using `internal_domain` from vars in 23e985c.

File: modules/aws/scalardl/scalardl.tf (hunk 2 above)
Comment (tei-k) [reply modules/aws/scalardl/scalardl.tf]: >Renaming of the variable (to blue_enable_dns_for_envoy or something)

This looks good.

- There are few fixes.
- It's easy to ssh in the case of disable. 🤔

Comment (feeblefakie) [reply modules/aws/scalardl/scalardl.tf:165]: I suggested, but `blue_enable_dns_for_envoy` is still a bit confusing.
How about `blue_discoverable_by_envoy` ?

Comment (tei-k) [reply modules/aws/scalardl/scalardl.tf:165]: 👌 Fixed to `blue_discoverable_by_envoy ` and `green_discoverable_by_envoy ` in 32378d6.

File: docs/BlueGreenDeploy.md [hunk 3]
```diff
@@ -14,39 +14,59 @@ This module manages two Scalar DL clusters, blue and green. At any given time on
...
+  blue_discoverable_by_envoy  = "true"
+  green_resource_count        = "3"
+  green_image_tag             = "2.1.0"
+  green_image_name            = "scalarlabs/scalar-ledger"
+  green_discoverable_by_envoy = "true"
 }
 ```
 
-* Remove Blue Cluster (Step 2)
+* Disable Blue Cluster DNS (Step 2)
```
Comment (feeblefakie) [inline docs/BlueGreenDeploy.md:46]:
```suggestion
* Make blue cluster not discoverable by Envoy (Step 2)
This makes all the requests from Envoy will go to green eventually.
```

File: docs/BlueGreenDeploy.md [hunk 4]
```diff
@@ -14,39 +14,59 @@ This module manages two Scalar DL clusters, blue and green. At any given time on
 [ [Azure example.tfvars](../examples/azure/scalardl/example.tfvars) ]
 [ [AWS example.tfvars](../examples/aws/scalardl/example.tfvars) ]
 
-* Green Cluster Active (Initial State)
+* Blue Cluster Active (Initial State)
```
Comment (feeblefakie) [inline docs/BlueGreenDeploy.md:17]:
```suggestion
* Blue cluster is in an active state (initial state)
```

File: docs/BlueGreenDeploy.md [hunk 5]
```diff
@@ -14,39 +14,59 @@ This module manages two Scalar DL clusters, blue and green. At any given time on
...
+  blue_image_name             = "scalarlabs/scalar-ledger"
+  blue_discoverable_by_envoy  = "true"
+  green_resource_count        = "0"
+  green_image_tag             = "2.0.1"
+  green_image_name            = "scalarlabs/scalar-ledger"
+  green_discoverable_by_envoy = "false"
 }
 ```
 
 * Deploy Green Cluster version 2.1.0 (Step 1)
```
Comment (feeblefakie) [inline docs/BlueGreenDeploy.md:32]:
```suggestion
* Deploy green cluster version 2.1.0 (Step 1)
```

File: docs/BlueGreenDeploy.md [hunk 6]
```diff
@@ -14,39 +14,59 @@ This module manages two Scalar DL clusters, blue and green. At any given time on
...
+  blue_image_name             = "scalarlabs/scalar-ledger"
+  blue_discoverable_by_envoy  = "false"
+  green_resource_count        = "3"
+  green_image_tag             = "2.1.0"
+  green_image_name            = "scalarlabs/scalar-ledger"
+  green_discoverable_by_envoy = "true"
+}
+```
+
+* Remove Blue Cluster (Step 3)
```
Comment (feeblefakie) [inline docs/BlueGreenDeploy.md:60]:
```suggestion
* Remove blue cluster (Step 3)
It should be done after making sure that requests from Envoy are not going to blue any more.
```

File: docs/BlueGreenDeploy.md [hunk 7]
```diff
@@ -14,39 +14,59 @@ This module manages two Scalar DL clusters, blue and green. At any given time on
...
-  blue_resource_count  = "0"
-  blue_image_tag       = "2.0.1"
-  blue_image_name      = "scalarlabs/scalar-ledger"
-  green_resource_count = "3"
-  green_image_tag      = "2.1.0"
-  green_image_name     = "scalarlabs/scalar-ledger"
+  blue_resource_count         = "3"
+  blue_image_tag              = "2.0.1"
+  blue_image_name             = "scalarlabs/scalar-ledger"
+  blue_discoverable_by_envoy  = "false"
```
Comment (feeblefakie) [inline docs/BlueGreenDeploy.md:52]:
```suggestion
  blue_discoverable_by_envoy  = "false" <- this is set to `false`
```

File: docs/BlueGreenDeploy.md [hunk 8]
```diff
@@ -14,39 +14,59 @@ This module manages two Scalar DL clusters, blue and green. At any given time on
...
+  green_image_tag             = "2.1.0"
+  green_image_name            = "scalarlabs/scalar-ledger"
+  green_discoverable_by_envoy = "true"
+}
+```
+
+* Remove Blue Cluster (Step 3)
+```
+scalardl = {
+  blue_resource_count         = "0"
```
Comment (feeblefakie) [inline docs/BlueGreenDeploy.md:63]:
```suggestion
  blue_resource_count         = "0" <- this is set to 0
```

**Comment by tei-k** [comment]
@feeblefakie 

Thanks!
Fixed as suggested. (22fb679)

**Comment by feeblefakie** [comment]
@ymorimo @penpyt

**Comment by penpyt** [comment]
LGTM, I followed the B/G docs on azure no timeout, no error, super smooth 👍

**Comment by feeblefakie** [comment]
@penpyt Please give approval instead of comment. 👍

File: docs/BlueGreenDeploy.md (hunk 8 above)
Comment (feeblefakie) [reply docs/BlueGreenDeploy.md:65]: @tei-k Forgot to apply ?

Comment (tei-k) [reply docs/BlueGreenDeploy.md]: 😱 Sorry! Added in d3223d0.

File: modules/universal/envoy/provision/envoy-tls.yaml.j2 [hunk 9]
```diff
@@ -45,7 +45,8 @@ static_resources:
                       route:
                           cluster: scalar-service
                     cors:
-                      allow_origin: ["*"]
+                      allow_origin_string_match:
+                      - prefix: "*"
```
Comment (ymorimo) [inline modules/universal/envoy/provision/envoy-tls.yaml.j2:49]:
```suggestion
                        - prefix: "*"
```

File: modules/universal/envoy/provision/envoy.yaml.j2 [hunk 10]
```diff
@@ -45,7 +45,8 @@ static_resources:
                       route:
                           cluster: scalar-service
                     cors:
-                      allow_origin: ["*"]
+                      allow_origin_string_match:
+                      - prefix: "*"
```
Comment (ymorimo) [inline modules/universal/envoy/provision/envoy.yaml.j2:49]:
```suggestion
                        - prefix: "*"
```

File: modules/universal/envoy/provision/envoy.yaml [hunk 11]
```diff
@@ -45,7 +45,8 @@ static_resources:
                       route:
                           cluster: scalar-service
                     cors:
-                      allow_origin: ["*"]
+                      allow_origin_string_match:
+                      - prefix: "*"
```
[the remark this answers is not in the document]
Comment (tei-k) [reply modules/universal/envoy/provision/envoy.yaml:49]: @ymorimo 

Thanks! 👍
I'm referring to the files below.
Since the test has been completed, I think it would be better do in other PR(yml refactoring?)...
What do you think?

https://github.com/grpc/grpc-web/blob/cc5cbf6e40903677333082909e30be1650e83428/net/grpc/gateway/examples/helloworld/envoy.yaml#L28

[the remark this answers is not in the document]
Comment (feeblefakie) [reply modules/universal/envoy/provision/envoy.yaml:49]: @tei-k Please do the yaml refactoring in another PR.
I'll merge it.
